utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/collection.scala

```diff
@@ -0,0 +1,248 @@
+import org.springframework.util.{NumberUtils => SpringNumberUtils}
+import pl.touk.nussknacker.engine.api.generics.{GenericFunctionTypingError, GenericType, TypingFunction}
+import pl.touk.nussknacker.engine.api.typed.supertype.NumberTypesPromotionStrategy.ForLargeNumbersOperation
+import pl.touk.nussknacker.engine.api.typed.typing
+import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypedClass, TypedObjectTypingResult, Unknown}
+import pl.touk.nussknacker.engine.api.{Documentation, HideToString, ParamName}
+
+import java.util.{Collections, Objects}
+import scala.collection.immutable.ListMap
+import scala.jdk.CollectionConverters._
+import scala.language.higherKinds
+import scala.reflect.ClassTag
+
+object collection extends HideToString {
+
+  @Documentation(description = "Concatenates two lists")
+  @GenericType(typingFunction = classOf[ListAdditionTyping])
+  def concat[T](@ParamName("list1") list1: java.util.List[T], @ParamName("list2") list2: java.util.List[T]): java.util.List[T] =
+    (list1.asScala.toList ++ list2.asScala).asJava
+
+  @Documentation(description = "Merges two maps. Values in the first map will be overwritten with values from the second map if keys are the same")
+  @GenericType(typingFunction = classOf[MapMergeTyping])
+  def merge[K, V](@ParamName("map1") map1: java.util.Map[K, V], @ParamName("map2") map2: java.util.Map[K, V]): java.util.Map[K, V] = {
+    val merged = new java.util.LinkedHashMap[K, V](map1)
+    merged.putAll(map2)
+    merged
+  }
+
+  @Documentation(description = "Returns the smallest element (elements must be comparable)")
+  @GenericType(typingFunction = classOf[ListElementTyping])
+  def min[T <: Comparable[T]](@ParamName("list") list: java.util.Collection[T]): T = Collections.min[T](list)
+
+  @Documentation(description = "Returns the largest element (elements must be comparable)")
+  @GenericType(typingFunction = classOf[ListElementTyping])
+  def max[T <: Comparable[T]](@ParamName("list") list: java.util.Collection[T]): T = Collections.max[T](list)
+
+  @Documentation(description = "Returns a slice of the list starting with start index (inclusive) and ending at stop index (exclusive)")
+  @GenericType(typingFunction = classOf[ListTyping])
+  def slice[T](@ParamName("list") list: java.util.Collection[T], @ParamName("start") start: Int, @ParamName("stop") stop: Int): java.util.List[T] =
+    list.asScala.slice(start, stop).toList.asJava
+
+  // TODO: This method currently has the following limitations:
+  // * it won't return the expected type for empty list
+  // * it works slower than it could with statically typed result
+  // To overcome this limitations we should supply returnType to this method
+  @Documentation(description = "Returns a sum of all elements")
+  @GenericType(typingFunction = classOf[ListElementTypingForSum])
+  def sum[T <: java.lang.Number](@ParamName("listOfNumbers") list: java.util.Collection[T]): T = {
+    val listType = Typed.fromInstance(list)
```

Comment:
It won't be a `List[Number]` or `List[Uknown]` anytime? If so, let's remove this and leave only:
```
    val targetType = 
      list.asScala.map(Typed.fromInstance(_)).reduce { (a, b) =>
        ForLargeNumbersOperation.promote(a, b)
      }
```
type computation method

Reply:
`Typed.fromInstance(list)` would go through list elements, so `listType` be a more detailed type. On the other hand, this is exactly what we do here:
```
list.asScala.map(Typed.fromInstance(_)).reduce { (a, b) =>
  ForLargeNumbersOperation.promote(a, b)
}
```
So the first loop is redundant. You are right, I've left only the latter computation method.